ENH: added functionality to handle scalars in `stats._chk_asarray`

aeklant opened the pull request:
See issue #4550 for details.

aeklant commented:
It seems I missed a blank line for PEP8 compliance which I already fixed.

What I'm not sure about is what should be the expected result for `spearmanr` and probably `sem` too. Should they raise a ValueError when receiving a scalar as input?

jaimefrio commented on scipy/stats/stats.py:
```diff
@@ -205,6 +205,9 @@
 
 
 def _chk_asarray(a, axis):
+    if len(np.array(a).shape) == 0:
```
This would typically be checked as `if np.array(a).ndim == 0:`, and similarly in the other two occurrences.

aeklant replied:
Thanks @jaimefrio. I made the adjustment. I am new to the community and trying to become a useful member so I quite value your input.

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -501,6 +505,11 @@ class TestCorrSpearmanr(TestCase):
+    def test_scalar(self):
+        y = stats.spearmanr(4.)
+        r = y[0]
+        assert(np.isnan(r))
```
This should be ``assert_``. Reason: ``assert`` statements get stripped when byte-compiling with ``-O``, so we don't use them for tests.

rgommers replied:
Also, I would not use the line above but write this in a more condensed form (``assert_(np.isnan(y[0]))``

rgommers replied:
and there are two more ``assert``s below to change

aeklant replied:
Ok, it makes sense. Change has been made.

rgommers commented:
I think that returning ``(nan, nan)`` for input to ``spearmanr``, which you test for and what it does at the moment, is fine. I would do an explicit check though, on the line below ``_chk_asarray``:

    if a.size <= 1:
        return np.nan, np.nan

And the same for ``b``. Can you also update your test to check that it works the same for ``b``?

rgommers commented:
And the same for ``sem``, returning ``nan`` is fine. That's what it does now for inputs ``[]`` and ``[1]``, but currently scalar input raises a confusing exception.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -205,6 +205,9 @@
 
 
 def _chk_asarray(a, axis):
+    if np.array(a).ndim == 0:
```
I know this is the line that @jaimefrio recommended, but it doesn't quite fit at this position. Array conversion is done twice now: here, and again in the ``asarray`` or ``ravel`` call below. It's more efficient to move the check to the end, then you can do:

    if a.ndim == 0:
        a = np.atleast_1d(a)

which is much faster.

rgommers replied:
and same applies to the 2-D version

aeklant replied:
Ok, done. Making a mental note for next time.

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -501,6 +505,15 @@ class TestCorrSpearmanr(TestCase):
+    def test_scalar(self):
+        y = stats.spearmanr(4.)
+        assert_(np.isnan(y[0]))
+        assert_(np.isnan(y[1]))
```
Maybe shorter: ``assert_(np.isnan(y).all())``

rgommers replied:
and the same below

aeklant replied:
I've made the change here. I decided to keep the long version for the `ttest_*` ones for consistency with the rest of those test cases.

aeklant commented on scipy/stats/stats.py:
```diff
@@ -2772,11 +2780,15 @@ def spearmanr(a, b=None, axis=0):
     if b is not None:
         b, axisout = _chk_asarray(b, axis)
+        if b.size <= 1:
+            return np.nan, np.nan
```
I added the size check here as well. I think it is good for clarity and consistency but it may be redundant because `np.corrcoef` raises an exception unless the sizes `a` and `b` are equal. What do you think @rgommers?

rgommers replied:
It's clearly documented that ``a`` and ``b`` need to be of the same length, so I'd prefer the current ``ValueError``.

Now ``stats.spearmanr([1, 2, 1], [8, 9])`` would raise an error but ``stats.spearmanr([1, 2, 1], 8) return ``nan, nan``.

rgommers replied:
So it seems that I missed that in my earlier comment, sorry about that.

aeklant commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2045,8 +2085,9 @@ def test_ttest_rel():
-        assert_equal(stats.ttest_ind(anan, np.zeros((2,2))),([0, np.nan], [1,np.nan]))
+        anan = np.array([[1, np.nan], [-1, 1]])
+        assert_equal(stats.ttest_rel(anan, np.zeros((2, 2))),
+                     ([0, np.nan], [1, np.nan]))
```
This seems to be an autocomplete error. The `ttest_ind` should not appear in `ttest_rel` tests cases and it is repeated a little later. Please let me know if I'm missing the point here and should rollback this change.

rgommers replied:
Yep, looks like a copy-paste error so your change is correct

aeklant commented:
I removed the check to `b.size` and also removed one of the scalar tests because they became redundant with that change.

I added two tests for different length input values for good measure since this was not tested.

rgommers commented:
Allright, Travis is happy and all comments are addressed, so in it goes. Thanks @aeklant

aeklant commented:
Happy to contribute :)
